src/stan/mcmc/hmc/base_hmc.hpp

```diff
@@ -141,6 +141,13 @@ class base_hmc : public base_mcmc {
     this->z_.ps_point::operator=(z_init);
   }
 
+  void metric_check(bool ok) {
```

Comment:
The error message is incorrect here.  Because the metric adaptation is convex the metric elements can become NaN only if the unconstrained states are NaN.  That in turn can happen from underflow or overflow in the gradient evaluation or even divergent numerical trajectories, which are both distinct  from the conditions of an improper posterior density where the gradient vanishes identically.

Reply:
In the example model given on the issue the _metric_ does not become NaN but _infinite_ which makes `accept_stat` and `stepsize` into `NaN`, and that happens at finite value of unconstrained state. (The square of a sufficiently large finite value is infinite.)
It's true that technically this can happen without the posterior being improper (and improper posterior does not require the gradient to vanish) but I figured that if the distribution is so wide that floating-point arithmetic fails then we might as well pretend it reaches all the way out to infinity.

Reply:
Yes, I am treating NaN/inf as the same for diagnostic purposes.

The problem with "improper" is that is can never be fully verified empirically.  The original warning was designed for flat posterior densities functions because of the propensity of early users to use uniform prior densities with nonidentified likelihood functions, where the flat posterior density identified the flat ledge that (was assumed to) roll all the way to infinity.   Most common improper prior densities are asymptotically flat and eventually encounter the same behavior.

The example here is different because the target density is worse than flat.  But we have no way of verifying if the problem will continue to infinity (and actually be improper) because the floating point falls apart and we don't have an interpretation like the above to motivate a more theoretical extrapolation.

We can avoid the ambiguity entirely by just being precise about the warning -- if the inverse metric elements are NaN then report that they NaN'd and note that should be caused by NaN states, and if they're infinite report that they're infinite and suggest that is probably caused by extremely large states.